presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java

```diff
@@ -0,0 +1,491 @@
+        List<LongStreamCheckpoint> dataCheckpoints = dataStream.getCheckpoints();
+        Optional<List<BooleanStreamCheckpoint>> presentCheckpoints = presentStream.getCheckpoints();
+        for (int i = 0; i < rowGroups.size(); i++) {
+            int groupId = i;
+            ColumnStatistics columnStatistics = rowGroups.get(groupId).getColumnStatistics();
+            LongStreamCheckpoint dataCheckpoint = dataCheckpoints.get(groupId);
+            Optional<BooleanStreamCheckpoint> presentCheckpoint = presentCheckpoints.map(checkpoints -> checkpoints.get(groupId));
+            List<Integer> positions = createSliceColumnPositionList(compressionParameters.getKind() != NONE, dataCheckpoint, presentCheckpoint);
+            rowGroupIndexes.add(new RowGroupIndex(positions, columnStatistics));
+        }
+
+        Slice slice = compressedMetadataWriter.writeRowIndexes(rowGroupIndexes.build());
+        Stream stream = new Stream(column, StreamKind.ROW_INDEX, slice.length(), false);
+        return ImmutableList.of(new StreamDataOutput(slice, stream));
+    }
+
+    private static List<Integer> createSliceColumnPositionList(
+            boolean compressed,
+            LongStreamCheckpoint dataCheckpoint,
+            Optional<BooleanStreamCheckpoint> presentCheckpoint)
+    {
+        ImmutableList.Builder<Integer> positionList = ImmutableList.builder();
+        presentCheckpoint.ifPresent(booleanStreamCheckpoint -> positionList.addAll(booleanStreamCheckpoint.toPositionList(compressed)));
+        positionList.addAll(dataCheckpoint.toPositionList(compressed));
+        return positionList.build();
+    }
+
+    @Override
+    public List<StreamDataOutput> getDataStreams()
+    {
+        checkState(closed);
+        if (directEncoded) {
+            return getDirectColumnWriter().getDataStreams();
+        }
+
+        // actually write data
+        ImmutableList.Builder<StreamDataOutput> outputDataStreams = ImmutableList.builder();
+        presentStream.getStreamDataOutput(column).ifPresent(outputDataStreams::add);
+        outputDataStreams.add(dataStream.getStreamDataOutput(column));
+        outputDataStreams.addAll(getDictionaryStreams(column));
+        return outputDataStreams.build();
+    }
+
+    @Override
+    public long getBufferedBytes()
+    {
+        checkState(!closed);
+        if (directEncoded) {
+            return getDirectColumnWriter().getBufferedBytes();
+        }
+        // for dictionary columns we report the data we expect to write to the output stream
+        return getIndexBytes() + getDictionaryBytes();
+    }
+
+    @Override
+    public long getRetainedBytes()
+    {
+        return getInstanceSize() +
+                rowGroupIndexes.sizeOf() +
+                dataStream.getRetainedBytes() +
+                presentStream.getRetainedBytes() +
+                getRetainedDictionaryBytes() +
+                (getDirectColumnWriter() == null ? 0 : getDirectColumnWriter().getRetainedBytes()) +
```

Comment:
Rather than summing up subclass's instance size, dictionary size, and direct column writer size. We just need one abstract function `getRetainedDictionaryBytes` and report the total subclass size through it.